### ISSUE TITLE
Comparison between int32_t fields incorrect below 2^23

### ISSUE THREAD

**Issue by fundudeone** [CE-Programming/toolchain#393]
Found another issue similar to my last one [url](https://github.com/CE-Programming/toolchain/issues/391). When I compare to int32_t fields on an object the result is incorrect below 2^23. The objects are passed by reference into the comparison if that changes anything. I'll admit this could be my own shoddy work but the 2^23 thing is a TI calling card.

Code that reproduces:
```c++
#include <stdint.h>
#include <debug.h>
#include <keypadc.h>

class fixed{
	//fixed point number class where I discovered issue. Cut to relevant content
	
	private:
		static constexpr fixed form(int32_t v) {fixed k; k.value = v; return k;}
	
	public:
		int32_t value = 0;
		
		constexpr fixed() = default; //default constructor
		
		constexpr fixed(const float f){
			//float constructor
			value = int32_t(f * float(1 << 16) + (f >= 0 ? 0.5 : - 0.5));
		}
		
		constexpr operator float() const{
			//cast to float
			return float(value) / float(1 << 16);
		}
				
		
	
};
//boolean expressions
constexpr bool operator == (const fixed& a, const fixed& b){
	return int32_t(a.value) == int32_t(b.value);
}
	
bool operator > (const fixed& a, const fixed& b){
		
	dbg_printf("==============================================\n");
	dbg_printf("fixed: %li > %li = %d\n", a.value, b.value, (a.value > b.value));
	dbg_printf("float: %f > %f = %d\n", float(a), float(b), (float(a) > float(b)));
	dbg_printf("==============================================\n");
	return int32_t(a.value) > int32_t(b.value);
}
		
bool operator < (const fixed& a, const fixed& b){
	dbg_printf("==============================================\n");
	dbg_printf("%li < %li = %d\n", a.value, b.value, (a.value < b.value));
	dbg_printf("%f < %f = %d\n", float(a), float(b), (float(a) < float(b)));
	dbg_printf("==============================================\n");
	return int32_t(a.value) < int32_t(b.value);
}
		
bool operator <= (const fixed& a, const fixed& b){
	return (int32_t(a.value) < int32_t(b.value) || int32_t(a.value) == int32_t(b.value));
}
		
bool operator >= (const fixed& a, const fixed& b){
	return (int32_t(a.value) > int32_t(b.value) || int32_t(a.value) == int32_t(b.value));
}


int main(void)
{
	dbg_printf("\nint32_t comparisons (working):\n");
	int32_t i1 = 3276800l; //less than 2^23
	int32_t i2 = 15663104l; //significantly greater than 2^23
	int32_t i3 = 8388609l; //1 greater than 2^23
	dbg_printf("==============================================\n");
	dbg_printf("%li > %li = %d\n", i1, i2, (i1 > i2));
	dbg_printf("==============================================\n");
	dbg_printf("==============================================\n");
	dbg_printf("%li > %li = %d\n", i3, i2, (i3 > i2));
	dbg_printf("==============================================\n\n");
	
	dbg_printf("int32_t comparisons as field on reference (not working):\n");
	fixed f1(50); //"Value" is less than 2^23
	fixed f2(239); //"Value" is signficantly greater than 2^23
	fixed f3;
	f3.value = 8388609; //"Value" is greater than 2^23
	bool test1 = f1 > f2;
	bool test2 = f3 > f2;
	
    return 0;
}
```

**Comment by jacobly0** [comment]
`__lcmps` produces the wrong sign flag when the upper bytes are equal.

**Comment by fundudeone** [comment]
ok thanks, better than my previous workaround of casting to float for each comparison

**Comment by jacobly0** [comment]
Fixed by 4e397b8c7c7c21719da1b451d08535b2228568a5.
